packages/muelu/src/Graph/StructuredAggregation/MueLu_AggregationStructuredAlgorithm_def.hpp

```diff
@@ -134,11 +138,184 @@ namespace MueLu {
+      } else {
+        rate = geoData->getCoarseningEndRate(0);
+      }
+      if(iRem > (rate / 2)) { ++iCoarse; }
+      if(coupled && (geoData->getStartGhostedCoarseNode(0)*geoData->getCoarseningRate(0)
+                     > geoData->getStartIndex(0))) { --iCoarse; }
+      jCoarse = jGhosted / geoData->getCoarseningRate(1);
+      jRem    = jGhosted % geoData->getCoarseningRate(1);
+      if(jGhosted - geoData->getOffset(1)
+         < geoData->getLocalFineNodesInDir(1) - geoData->getCoarseningEndRate(1)) {
+        rate = geoData->getCoarseningRate(1);
+      } else {
+        rate = geoData->getCoarseningEndRate(1);
+      }
+      if(jRem > (rate / 2)) { ++jCoarse; }
+      if(coupled && (geoData->getStartGhostedCoarseNode(1)*geoData->getCoarseningRate(1)
+                     > geoData->getStartIndex(1))) { --jCoarse; }
+      kCoarse = kGhosted / geoData->getCoarseningRate(2);
+      kRem    = kGhosted % geoData->getCoarseningRate(2);
+      if(kGhosted - geoData->getOffset(2)
+         < geoData->getLocalFineNodesInDir(2) - geoData->getCoarseningEndRate(2)) {
+        rate = geoData->getCoarseningRate(2);
+      } else {
+        rate = geoData->getCoarseningEndRate(2);
+      }
+      if(kRem > (rate / 2)) { ++kCoarse; }
+      if(coupled && (geoData->getStartGhostedCoarseNode(2)*geoData->getCoarseningRate(2)
+                     > geoData->getStartIndex(2))) { --kCoarse; }
+      geoData->getCoarseNodeGhostedLID(iCoarse, jCoarse, kCoarse, ghostedCoarseNodeCoarseLID);
+
+      // Fill Graph
+      // Check if Fine node lies on Coarse Node
+      iIsCoarse = false; jIsCoarse = false; kIsCoarse = false;
+      if(iRem == 0)
+          iIsCoarse = true;
+      if(jRem == 0)
+          jIsCoarse = true;
+      if(kRem == 0)
+          kIsCoarse = true;
+      if(coupled){
+        if(iGhosted - geoData->getOffset(0) == geoData->getLocalFineNodesInDir(0)-1 && geoData->getMeshEdge(0*2+1) )
+          iIsCoarse = true;
+        if(jGhosted - geoData->getOffset(1) == geoData->getLocalFineNodesInDir(1)-1 && geoData->getMeshEdge(1*2+1) )
+          jIsCoarse = true;
+        if(kGhosted - geoData->getOffset(2) == geoData->getLocalFineNodesInDir(2)-1 && geoData->getMeshEdge(2*2+1) )
+          kIsCoarse = true;
+      } else {
+        if(iGhosted - geoData->getOffset(0) == geoData->getLocalFineNodesInDir(0)-1)
+            iIsCoarse = true;
+        if(jGhosted - geoData->getOffset(1) == geoData->getLocalFineNodesInDir(1)-1)
+            jIsCoarse = true;
+        if(kGhosted - geoData->getOffset(2) == geoData->getLocalFineNodesInDir(2)-1)
+            kIsCoarse = true;
+      }
+
+      if(iIsCoarse && jIsCoarse && kIsCoarse){ // Fine node lies on Coarse node
+        geoData->getCoarseNodeGhostedLID(iCoarse,  jCoarse,  kCoarse,  colIndex[rowPtr[nodeIdx] + 0]);
+        nnzOnRow[nodeIdx] = 1;
+        rowPtr[nodeIdx+1] = rowPtr[nodeIdx] + 1;
+      } else {
+        nnzOnRow[nodeIdx] = std::pow(2.0,dim);
```

Comment:
You are implicitely converting a `double` into a `size_t`, please use `Teuchos::as<size_t>(val)` instead.

Reply:
It would be better to use left shift so that everything stays an integer -- `std::pow` doesn't promise anything about accuracy.  (Integer input to `std::pow` get promoted to `double` as of C++11.)

Reply:
I am a little bit worried about the readability of the code if we use bit operations, it might be easier for later maintenance to simply use `Teuchos::as<size_t>` on the return value of `std::pow()` even though this might be a little more expensive?

Reply:
@lucbv Sorry to poke my nose into your code :-) I'm just worried that `floor(std::pow(2.0,x))` might not be a power of 2.